Hot Fix - BuffMaintain Speculation Fix

Alium58 opened the pull request:
# Description

BuffMaintain when speculating would assume mall purchases would work. Which is fine, however we shouldn't even attempt to buy when in hardcore or if it is a quest item.

This fixes speculation thinking it can get buffs when it can not. Also fixes attempting to buy things and failing when not ran speculatively. 

Discovered when #1073 bought the ML hat when I couldn't max out -com or +com. 

## How Has This Been Tested?

Ash calls while in run

## Checklist:

- [x] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have based by pull request against the [master branch](https://github.com/Loathing-Associates-Scripting-Society/autoscend/tree/master) or have a good reason not to.

quarklikeadork commented:
only assume buy works in one of 2 cases? available in meat npc store or available in coinmaster store
npc_price(it) != 0
else
is_accessible(it.seller)

taltamir commented:
> only assume buy works in one of 2 cases? available in meat npc store or available in coinmaster store npc_price(it) != 0 else is_accessible(it.seller)

this looks good

quarklikeadork commented:
and what the latter option needs to check:
item.seller.available_tokens
VS sell_price(item.seller,item)

Alium58 commented:
Thanks for the ideas! Updated it to attempt to buy from NPC with meat, then NPC with coins, then mall. Also updated mall buying to respect meatReserve(). Working in HC TT run. Haven't tested in softcore

taltamir commented:
a couple of things:
1. I am not too comfortable with buying from token stores like this. some tokens are really valuable.

2. I get the feeling most of this code should be spun off to its own function which buffMaintain would then call. something like `auto_buy` instead.
Or maybe enhance an existing functions such as `buyUpTo` or `acquireOrPull`.
Although the latter needs to be repaired as it uses `retrieve_item` which is undesirable due to it triggering an error when it fails which can cause problems
also it uses pullXWhenHaveY which is depreciated due to nerfday and needs to be replaced everywhere in autoscend.

quarklikeadork commented:
> 1. I am not too comfortable with buying from token stores like this. some tokens are really valuable. what are even situations where we need to buy token items?

effects like Gummed Shoes would be from token store
since really valuable  tokens are probably not obtained by autoscend in run could just block coinmasters when can_interact

Alium58 commented:
Ran out of time to do much tonight. Looked closer at original code and it did not buy from coinmasters, I thought it did. Since it didn't originally, I took that out of this change.

Agree improvements to acquiring is needed. I think that is worth a new PR to resolve. Goal for this one is to have buffMaintain speculate properly and it can buy the -com hat
